pysph/sph/wc/density_correction.py

```diff
@@ -32,3 +111,140 @@ def initialize(self, d_idx, d_rho):
 
     def loop(self, s_idx, d_idx, s_m, d_tmp_w, d_rho, WIJ):
         d_rho[d_idx] += WIJ * s_m[s_idx] / d_tmp_w[d_idx]
+
+
+class MLSFirstOrderPreStep2D(Equation):
+    r"""**Moving Least Squares density reinitialization**
+    This is a first order density reinitialization
+
+    .. math::
+            W_{ab}^{MLS} = \beta\left(\mathbf{r_{a}}\right)\cdot\left(
+            \mathbf{r}_a - \mathbf{r}_b\right)W_{ab}
+
+    .. math::
+            \beta\left(\mathbf{r_{a}}\right) = A^{-1}
+            \left[1 0 0\right]^{T}
+
+    where
+
+    .. math::
+            A = \sum_{b}W_{ab}\tilde{A}\frac{m_{b}}{\rho_{b}}
+    References
+    ----------
+    .. [Dilts, 1999] Dilts, G. A. Moving-Least-Squares-Particle
+        Hydrodynamics - I. Consistency and stability,
+        Int. J. Numer. Meth. Engng, 1999.
+    """
+
+    def _get_helpers_(self):
+        return [gj_Solve]
+
+    def initialize(self, d_idx, d_a):
+        n = declare('int')
+        n = 3
+        i = declare('int')
+        for i in range(n * n):
+            d_a[n * n * d_idx + i] = 0.0
+
+    def loop(self, s_idx, d_idx, d_a, s_m, s_rho, WIJ, XIJ):
+        n = declare('int')
+        n = 3
+        i = declare('int')
+        j = declare('int')
+        for i in range(n):
+            for j in range(n):
+                fac1 = 1.0 if i == 0 else XIJ[i - 1]
+                fac2 = 1.0 if j == 0 else XIJ[j - 1]
+                d_a[n * n * d_idx + n * i + j] += fac1 * \
+                    fac2 * WIJ * s_m[s_idx] / s_rho[s_idx]
+
+    def post_loop(self, d_idx, s_idx, d_a, d_b):
+        a = declare('matrix((9, ))')
+        i = declare('int')
+        j = declare('int')
+        n = declare('int')
+        n = 3
+        for i in range(n):
+            for j in range(n):
+                a[n * i + j] = d_a[n * n * d_idx + n * i + j]
+        res = declare('matrix((3,))')
+        res[0] = 1.0
+        for i in range(1, n):
+            res[i] = 0.0
+        gj_Solve(a, [1., 0., 0.], n, res)
+        d_b[n * d_idx] = res[0]
+        d_b[n * d_idx + 1] = res[1]
+        d_b[n * d_idx + 2] = res[2]
+
+
+class MLSFirstOrderPreStep3D(Equation):
+
+    def _get_helpers_(self):
+        return [gj_Solve]
+
+    def initialize(self, d_idx, d_a):
+        n = declare('int')
+        n = 4
+        i = declare('int')
+        for i in range(n * n):
+            d_a[n * n * d_idx + i] = 0.0
+
+    def loop(self, s_idx, d_idx, d_a, s_m, s_rho, WIJ, XIJ):
+        n = declare('int')
+        n = 4
+        i = declare('int')
+        j = declare('int')
+        for i in range(n):
+            for j in range(n):
+                fac1 = 1.0 if i == 0 else XIJ[i - 1]
```

Comment:
I don't think the ternary operator will work on the GPU.  This should be easy to support but isn't currently.

Reply:
Also move the conditional outside the for j loop as it doesn't need to be evaluated 9 times and only 3 since it is independent of j.

Reply:
Do this for all similar cases in your code please.